test/karma/test-app/util.ts

```diff
@@ -141,52 +236,93 @@ export function setupDomTests(document: Document) {
-            const script = document.createElement('script') as HTMLScriptElement;
-            if (tmpScripts[i].src) {
-              script.src = tmpScripts[i].src;
-            }
-            if (tmpScripts[i].hasAttribute('nomodule')) {
-              script.setAttribute('nomodule', '');
-            }
-            if (tmpScripts[i].hasAttribute('type')) {
-              script.setAttribute('type', tmpScripts[i].getAttribute('type')!);
+          /**
+           * Re-generate script tags that are embedded in the loaded HTML file.
+           *
+           * Doing so allows JS files to be loaded (via script tags), when the HTML is served, without having to configure
+           * Karma to load the JS explicitly. This is done by adding the host/port combination to existing `src`
+           * attributes.
+           *
+           * Before:
+           * ```html
+           * <script type="module" src="/index.6127a5ed.js"></script>
+           * ```
+           *
+           * After:
+           * ```html
+           * <script src="http://localhost:9876/index.547a265b.js" type="module"></script>
+           * ```
+           */
+          const parseAndRebuildScriptTags = () => {
+            const tempScripts: NodeListOf<HTMLScriptElement> = testElement.querySelectorAll('script');
+            for (let i = 0; i < tempScripts.length; i++) {
+              const script: HTMLScriptElement = document.createElement('script');
+              if (tempScripts[i].src) {
+                script.src = tempScripts[i].src;
+              }
+              if (tempScripts[i].hasAttribute('nomodule')) {
+                script.setAttribute('nomodule', '');
+              }
+              if (tempScripts[i].hasAttribute('type')) {
+                const typeAttribute = tempScripts[i].getAttribute('type');
+                if (typeof typeAttribute === 'string') {
+                  // older DOM implementations would return an empty string to designate `null`
+                  // here, we interpret the empty string to be a valid value
+                  script.setAttribute('type', typeAttribute);
+                }
+              }
+              script.innerHTML = tempScripts[i].innerHTML;
+
+              if (tempScripts[i].parentNode) {
+                // the scripts were found by querying a common parent node, which _should_ still exist
+                tempScripts[i].parentNode!.insertBefore(script, tempScripts[i]);
+                tempScripts[i].parentNode!.removeChild(tempScripts[i]);
+              } else {
+                // if for some reason the parent node no longer exists, something's manipulated it while we were parsing
+                // the script tags. this can lead to undesirable & hard to debug behavior, fail.
+                reject('the parent node for script tags no longer exists. exiting.');
+              }
             }
-            script.innerHTML = tmpScripts[i].innerHTML;
+          };
```

Comment:
ok so as far as I can tell this function is basically intended to find scripts and convert their `src` attrs from relative to absolute URLs, taking advantage of the fact that a `<script>` tag that has a relative URL set as its `src` attr will print the full absolute URL when that attr is accessed.

One question I have is, is it necessary to create new script tags to do this? Just messing around in the dev console for a minute I was able to accomplish this by doing something like

```js
const script = document.createElement("script")
script.src = "/foo/bar/baz"
script.src = script.src
```

then the actual set value for `src` is the absolute URL w/ the domain and whatnot. Maybe there's a reason that doesn't work for us here, or some other reason why we want to swap the scripts out with new ones?

Reply:
I _think_ so, I haven't gotten Karma to serve the files correctly otherwise. I don't know for sure this is the only way to get Karma to serve the files (bc I was scratching my head on this loop for a minute), I just haven't gotten it to work without it

Reply:
hmm maybe something to do with creating a new script with a well-formed `src` attr vs modifying one that already exists...well if it works it works!